### PR TITLE
Use hit object time for timeline drag selection instead of relying on blueprint

### LINKED ISSUE #16410: Editor skips objects when drag selecting too fast in the timeline
### Discussed in https://github.com/ppy/osu/discussions/16398

<div type='discussions-op-text'>

<sup>Originally posted by **NotGumballer91** January 10, 2022</sup>
As the title says.
Drag selecting slowly is fine.

https://user-images.githubusercontent.com/64581009/148749267-427cd357-1d7e-43ca-91e8-b71b2b01a2a8.mp4

Could be related to #16356 as well.</div>

### LINKED ISSUE #20279: Placing lots of stacked notes, selecting them via timeline does not select all objects
### Type

Game behaviour

### Bug description

From #18077 
Selecting a stack of notes in the timeline only selects the top 10 objects instead of all of the objects on that stack. You would expect all of the notes to be selected so you could quickly delete the whole stack all at once.

### Screenshots or videos

https://user-images.githubusercontent.com/17460441/189781714-cdc4ef03-3aae-4fe0-b7ab-98a042ddd5e6.mp4

### Version

https://github.com/ppy/osu/commit/00c9763d67f80de1673003a45eadbf8b9bfc23a9

### Logs

[runtime.log](https://github.com/ppy/osu/files/9552580/runtime.log)

### PR DESCRIPTION
**Pull request by ekrctb**
- Fix #16410
- FIx #20279

`AllowDeselectionDuringDrag` behavior (not deselecting while clock is running) is removed for timeline.
Note that selecting a large number of hit objects is still very slow (#17712?), because all DHOs must be added and also `AddBlueprintFor` has quadratic behaviors.

### CODE REVIEW

**Comment by ekrctb** [comment]
Also, I have a plan to implement a similar time-based drag box for scrolling rulesets to resolve <https://github.com/ppy/osu/discussions/18148>.

File: osu.Game/Screens/Edit/EditorBeatmap.cs [hunk 1]
```diff
@@ -352,6 +352,8 @@ protected override void UpdateState()
             var updates = batchPendingUpdates.ToArray();
             batchPendingUpdates.Clear();
 
+            foreach (var h in deletes) SelectedHitObjects.Remove(h);
+
             foreach (var h in deletes) HitObjectRemoved?.Invoke(h);
```
Comment (peppy) [inline osu.Game/Screens/Edit/EditorBeatmap.cs:357]: Is there a reason for doing the looping over `deletes` in two passes? More curious than anything.

**Comment by peppy** [comment]
I hit a few weird issues while testing this in extreme conditions:

- Hit `Z` to seek to start of beatmap
- Start drag using mouse
- While dragging, hit `V` to seek to end of beatmap
- Not all hitobjects are selected properly

In addition, the selection box can get stuck in this state.

https://user-images.githubusercontent.com/191335/194209893-efb324e3-72b4-42e7-8578-c8f27b089234.mp4

Haven't looked into it in detail, but may be something to do with lifetimes / pooling / masking.

**Comment by ekrctb** [comment]
Actually all hit objects are selected, but not all corresponding blueprints are added. For now, it should be fixed by adding all blueprints.
Though for the future we should consider whether all blueprints must be added or not (not sure about design) because it is too slow.

**Comment by ekrctb** [comment]
<https://github.com/ppy/osu/pull/20586/commits/0d448e6cc898e181ab65f31280e3be15ee761f42> (fix items without blueprints are not deselected) is not strictly needed because such case (selected hit objects without blueprints) is fixed by <https://github.com/ppy/osu/pull/20586/commits/29cc55463268c40b688d1125ea7b69886c046135> but I think it is still good because there may be other edge cases and such case occurs.

**Comment by peppy** [comment]
> Though for the future we should consider whether all blueprints must be added or not (not sure about design) because it is too slow.

Yeah, I'm not too sure on the best way to handle this just yet. Because there is a general expectation of what to visually expect from the stable editor. You still want the user to be aware of what they have selected, but there needs to be a way to do so without the overhead of 100% of blueprints.

I've thought about this over the last few years and don't have anything yet 😅.

File: osu.Game/Screens/Edit/Compose/Components/EditorBlueprintContainer.cs [hunk 2]
```diff
@@ -133,8 +131,8 @@ protected override bool OnDoubleClick(DoubleClickEvent e)
         protected override void SelectAll()
         {
             Composer.Playfield.KeepAllAlive();
-
-            base.SelectAll();
+            SelectedItems.Clear();
+            SelectedItems.AddRange(Beatmap.HitObjects);
```
Comment (peppy) [inline osu.Game/Screens/Edit/Compose/Components/EditorBlueprintContainer.cs:135]: I'm not sure about forcing a full re-selection here. At least while the operation has some overhead, I'd hope a second select all call is a no-op..

**Review by peppy — APPROVED** [review]

File: osu.Game/Screens/Edit/EditorBeatmap.cs (hunk 1 above)
Comment (ekrctb) [reply osu.Game/Screens/Edit/EditorBeatmap.cs:357]: (I missed this comment) no specific reason, just styled like other one-line `foreach`es.

**Pull request merged**